src/cluster_asm.c

```diff
@@ -271,6 +283,68 @@ static clusterNode *validateImportSlotRanges(slotRangeArray *slot_ranges, sds *e
     return *err ? NULL : source;
 }
 
+/* Returns 1 if a migrate operation is in progress, 0 otherwise. */
+int asmMigrateInProgress(void) {
+    listIter li;
+    listNode *ln;
+
+    if (!server.cluster_enabled || listLength(asmManager->tasks) == 0)
+        return 0;
+
+    listRewind(asmManager->tasks, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        asmTask *task = listNodeValue(ln);
+        if (task->operation == ASM_MIGRATE) return 1;
+    }
+    return 0;
+}
+
+/* Returns 1 if the task is in a state where it can receive replication stream
+*  for the slot range, 0 otherwise. */
+int asmCanFeedMigrationClient(asmTask *task) {
+    return task->operation == ASM_MIGRATE &&
+             (task->state == ASM_SEND_BULK_AND_STREAM ||
+              task->state == ASM_WAIT_PAUSE_WRITE);
+}
+
+/* Feed the migration client with the replication stream for the slot range. */
+void asmFeedMigrationClient(robj **argv, int argc) {
+    listIter li;
+    listNode *ln;
+    asmTask *task = NULL;
+
+    if (server.cluster_enabled == 0) return;
+
+    /* Quick check if there is a migrate task in progress. */
+    listRewind(asmManager->tasks, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        task = listNodeValue(ln);
+        if (task->operation == ASM_MIGRATE)
+            break;
+    }
+    if (!task) return;
+
+    /* Check if the command belongs to the slot range. */
+    struct redisCommand *cmd = lookupCommandBySds(argv[0]->ptr);
+    serverAssert(cmd);
+
+    int slot = getSlotFromCommand(cmd, argv, argc);
+    /* If the command does not have keys, or has crossslot keys, skip it.
+     * TODO: revisit this to see if we are okay with this. */
+    if (slot == GETSLOT_NOKEYS || slot == GETSLOT_CROSSSLOT) return;
+
+    /* Find the task that owns the slot. */
+    slotRange sr = {slot, slot};
+    task = lookupAsmTaskBySlotRange(&sr);
```

Comment:
we find the task , maybe we just check if the current task has this slot?

Reply:
it does exactly that? no? It returns NULL if the task does not have this slot.

Reply:
i mean we don't need iterate the asmManager->task agin, just check the task->slot_ranges

Reply:
I see, as we assume single task, no need to iterate anymore. I pushed something for it.